ForumProject/projects/models.py

```diff
@@ -2,56 +2,181 @@
+
+            # Compare attributes to identify changes
+            changes = []
+            for field in self._meta.fields:
+                field_name = field.name
+                if getattr(original, field_name) != getattr(self, field_name):
+                    changes.append(
+                        (field_name, getattr(original, field_name), getattr(self, field_name))
+                    )
+
+            if changes:
+                # Prepare the previous and modified states
+                previous_state = ', '.join([f'{field}: {old_value}' for field, old_value, _ in changes])
+                modified_state = ', '.join([f'{field}: {new_value}' for field, _, new_value in changes])
+        else:
+            action = 'Created Project'
+            previous_state = 'n/a'
+            modified_state = f'New Project, id: {self.pk}, name: {self.name}'
+
+        super().save(*args, **kwargs)
+        
+        # Create a new ProjectLog
+        current_date = datetime.now().date()
+        current_time = datetime.now().time()
+
+        ProjectLog.objects.create(
+            project=self,
+            project_title=self.name,
+            change_date=current_date,
+            change_time=current_time,
+            user_id=1,  # This is a placeholder
+            action=action,
+            previous_state=previous_state[:ProjectLog._meta.get_field('previous_state').max_length],
+            modified_state=modified_state[:ProjectLog._meta.get_field('modified_state').max_length]
+            )
+    
     def __str__(self):
         """
         Return the name of the project.
         
         Returns:
             str: The name of the project.
         """
-        return self.project_name
+        return self.name
+
+class ProjectFiles(models.Model):
 
+    def _generate_upload_path(self, filename):
+        '''
+        The function creates a valid path for each Project's documentation upload
+        and ensures renaming of the file being uploaded if its name is not unique in 
+        the selected folder.
+        '''
+        # Replace special characters with underscores and truncate to 20 characters
+        startup_slug = slugify(self.project.startup.startup_name)[:20]
+        project_slug = slugify(self.project.name)[:20]
+
+        # Determine the folder path
+        folder_path = os.path.join('media', f'projects/{startup_slug}/{project_slug}/')
+
+        # Ensure the folder exists
+        os.makedirs(folder_path, exist_ok=True)
+
+        # Replace spaces in the filename with underscores
+        filename = filename.replace(" ", "_")
+        
+        # Create a unique file path
+        full_path = os.path.join(folder_path, filename)
+
+        # Check for existing files with the same name and create a unique name
+        if os.path.exists(full_path):
+            base, extension = os.path.splitext(filename)
+            counter = 1
+            while os.path.exists(full_path):
```

Comment:
Consider adding error handling in case the file operations (like `os.path.exists`) fail

Reply:
Done